crates/trigger-http/src/lib.rs

```diff
@@ -1000,3 +1007,212 @@ mod tests {
+        }
+    } else {
+        let tcp_stream = TokioIo::new(tcp_stream);
+        let (sender, conn) = timeout(
+            connect_timeout,
+            // TODO: we should plumb the builder through the http context, and use it here
+            hyper::client::conn::http1::handshake(tcp_stream),
+        )
+        .await
+        .map_err(|_| types::ErrorCode::ConnectionTimeout)?
+        .map_err(hyper_request_error)?;
+
+        let worker = wasmtime_wasi::runtime::spawn(async move {
+            match conn.await {
+                Ok(()) => {}
+                // TODO: same as above, shouldn't throw this error away.
+                Err(e) => tracing::warn!("dropping error {e}"),
+            }
+        });
+
+        (sender, worker)
+    };
+
+    // at this point, the request contains the scheme and the authority, but
+    // the http packet should only include those if addressing a proxy, so
+    // remove them here, since SendRequest::send_request does not do it for us
+    *request.uri_mut() = http::Uri::builder()
+        .path_and_query(
+            request
+                .uri()
+                .path_and_query()
+                .map(|p| p.as_str())
+                .unwrap_or("/"),
+        )
+        .build()
+        .expect("comes from valid request");
+
+    let resp = timeout(first_byte_timeout, sender.send_request(request))
+        .await
+        .map_err(|_| types::ErrorCode::ConnectionReadTimeout)?
+        .map_err(hyper_request_error)?
+        .map(|body| body.map_err(hyper_request_error).boxed());
+
+    Ok(wasmtime_wasi_http::types::IncomingResponse {
+        resp,
+        worker: Some(worker),
+        between_bytes_timeout,
+    })
+}
+
+fn get_client_tls_config_for_authority(
+    authority: &String,
+    client_tls_opts: Option<HashMap<String, ParsedClientTlsOpts>>,
+) -> rustls::ClientConfig {
+    // derived from https://github.com/tokio-rs/tls/blob/master/tokio-rustls/examples/client/src/main.rs
+    let mut root_cert_store = rustls::RootCertStore {
+        roots: webpki_roots::TLS_SERVER_ROOTS.into(),
+    };
+
+    let client_tls_opts = match client_tls_opts {
+        Some(opts) => opts,
+        _ => {
+            return rustls::ClientConfig::builder()
+                .with_root_certificates(root_cert_store)
+                .with_no_client_auth();
+        }
+    };
+
+    let client_tls_opts_for_host = match client_tls_opts.get(authority) {
+        Some(opts) => opts,
+        _ => {
+            return rustls::ClientConfig::builder()
+                .with_root_certificates(root_cert_store)
+                .with_no_client_auth();
+        }
+    };
+
+    if let Some(custom_root_ca) = &client_tls_opts_for_host.custom_root_ca {
+        for cer in custom_root_ca {
+            match root_cert_store.add(cer.to_owned()) {
+                Ok(_) => {}
+                Err(e) => {
+                    tracing::warn!("failed to add custom cert to root_cert_store. error: {}", e)
```

Comment:
Is there a use case where we don't want this to be a fatal error?

Reply:
rustls has two implementations of adding certificates to a certificate store:

1. Adding a short list of known certs: https://docs.rs/rustls/latest/rustls/struct.RootCertStore.html#method.add
2. Adding a long list of certs, such as an OS root cert store: https://docs.rs/rustls/latest/rustls/struct.RootCertStore.html#method.add_parsable_certificates

The former is useful when adding a small set of certs that are known to the user, and the caller would want to know if there was an issue adding those certificates to the store.

The latter is more useful when you are adding a large list of certificates to your store. This is because large collections of root certificates often include ancient or syntactically invalid certificates. `add_parsable_certificates` adds as many certificates to the store as possible in a best-effort fashion.

The intended use case for this feature leans more towards the former than the latter.

Reply:
👍  this is done.